idx/memory/memory.go

```diff
@@ -0,0 +1,523 @@
+package memory
+
+import (
+	"flag"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxOk             met.Count
+	idxFail           met.Count
+	idxAddDuration    met.Timer
+	idxGetDuration    met.Timer
+	idxListDuration   met.Timer
+	idxFindDuration   met.Timer
+	idxDeleteDuration met.Timer
+
+	Enabled bool
+)
+
+func ConfigSetup() {
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.RWMutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
+}
+
+func New() *MemoryIdx {
+	return &MemoryIdx{
+		DefById: make(map[string]*schema.MetricDefinition),
+		Tree:    make(map[int]*Tree),
+	}
+}
+
+func (m *MemoryIdx) Init(stats met.Backend) error {
+	idxOk = stats.NewCount("idx.ok")
+	idxFail = stats.NewCount("idx.fail")
+	idxAddDuration = stats.NewTimer("idx.add_duration", 0)
+	idxGetDuration = stats.NewTimer("idx.get_duration", 0)
+	idxListDuration = stats.NewTimer("idx.list_duration", 0)
+	idxFindDuration = stats.NewTimer("idx.find_duration", 0)
+	idxDeleteDuration = stats.NewTimer("idx.delete_duration", 0)
+	return nil
+}
+
+func (m *MemoryIdx) Stop() {
+	return
+}
+
+func (m *MemoryIdx) Add(data *schema.MetricData) {
+	pre := time.Now()
+	m.Lock()
+	defer m.Unlock()
+	existing, ok := m.DefById[data.Id]
+	if ok {
+		log.Debug("metricDef with id %s already in index.", data.Id)
+		existing.LastUpdate = data.Time
+		idxOk.Inc(1)
+		idxAddDuration.Value(time.Since(pre))
+		return
+	}
+
+	def := schema.MetricDefinitionFromMetricData(data)
+	m.add(def)
+	idxAddDuration.Value(time.Since(pre))
+}
+
+func (m *MemoryIdx) Load(defs []schema.MetricDefinition) {
```

Comment:
the purpose and assumptions need to be clarified here.
* who will call this function (any persistent index that extends memory index?) and why
* why does it not trigger any metrics. does it assume the caller has their own?
* must all input defs be not existent in the index yet, what happens if they are?